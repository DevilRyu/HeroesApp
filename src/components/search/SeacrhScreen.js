import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import HeroCard from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SeacrhScreen = ({ history }) => {

    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);

    const initialForm = {
        searchText: q
    };

    const [formValues, handleInputChange] = useForm(initialForm);

    const { searchText } = formValues;

    //solo si quiero busquedas con enter y el boton
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q])

    //const heroesFiltered = getHeroesByName(searchText);

    const handleSearch = (e) => {

        e.preventDefault();
        history.push(`?q=${searchText}`);

    }

    return (
        <div>
            <div className="row">
                <div className="col-12 col-md-5">
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            placeholder="Find your hero"
                            className="form-control"
                            name="searchText"
                            value={searchText}
                            onChange={handleInputChange}
                            autoComplete="off"
                        />
                        <button
                            type="submit"
                            className="btn mt-2 w-100 btn-outline-primary"
                        >
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-12 col-md-7">
                    <h4>Results</h4>
                    <hr />
                    {
                        (q === '') 
                            &&
                        < div className="alert alert-info">
                            Search a hero
                        </div>
                    }
                    {
                        (q !== '' && heroesFiltered.length === 0) 
                            &&
                        < div className="alert alert-danger">
                            There is no a hero with {q}
                        </div>
                    }
                    {
                        heroesFiltered.map(hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>
            </div>
        </div >
    )
}
