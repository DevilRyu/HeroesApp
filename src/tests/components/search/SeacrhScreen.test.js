import React from 'react';
import { mount } from 'enzyme';
import { SeacrhScreen } from '../../../components/search/SeacrhScreen';
import { MemoryRouter, Route } from 'react-router';

describe('Pruebas en <SeacrhScreen />', () => {

    test('debe de mostrarse correctamente con valores por defecto', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/HeroesApp/search']}>
                <Route path="/HeroesApp/search" component={SeacrhScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');

    });

    test('debe de mostrar a Batman y el input con el valor del query string', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/HeroesApp/search?q=batman']}>
                <Route path="/HeroesApp/search" component={SeacrhScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();

    });

    test('debe de un error si no se encuentra el hero', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/HeroesApp/search?q=batman123']}>
                <Route path="/HeroesApp/search" component={SeacrhScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert').exists()).toBe(true);
        expect(wrapper).toMatchSnapshot();

    });

    test('debe de llamar el push del history', () => {

        const historyMock = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/HeroesApp/search?q=batman123']}>
                <Route
                    path="/HeroesApp/search"
                    component={() => <SeacrhScreen history={historyMock}/>}
                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change',{
            target:{
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(historyMock.push).toHaveBeenCalledWith('?q=batman');

    });

});