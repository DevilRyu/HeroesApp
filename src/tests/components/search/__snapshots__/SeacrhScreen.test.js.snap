// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SeacrhScreen /> debe de mostrar a Batman y el input con el valor del query string 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-12 col-md-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-2 w-100 btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-12 col-md-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card"
        >
          <img
            alt="Batman"
            className="card-img-top"
            src="/../assets/heroes/dc-batman.jpg"
          />
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <p
              className="card-text"
            >
              <strong>
                Alter Ego:
              </strong>
               
              Bruce Wayne
            </p>
            <p
              className="card-text"
            >
              <strong>
                Characters:
              </strong>
                
              Bruce Wayne
               
            </p>
            <a
              href="/HeroesApp/hero/dc-batman"
              onClick={[Function]}
            >
              MÃ¡s..
            </a>
          </div>
          <div
            className="card-footer"
          >
            <small
              className="text-muted"
            >
              <strong>
                Firt Appearance:
              </strong>
               
              Detective Comics #27
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SeacrhScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-12 col-md-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-2 w-100 btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-12 col-md-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SeacrhScreen /> debe de un error si no se encuentra el hero 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-12 col-md-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman123"
        />
        <button
          className="btn mt-2 w-100 btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-12 col-md-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no a hero with 
        batman123
      </div>
    </div>
  </div>
</div>
`;
